pkg/master/controller.go

```diff
@@ -0,0 +1,218 @@
+	ServiceIPRegistry service.IPRegistry
+	EndpointRegistry  endpoint.Registry
+	PortalNet         *net.IPNet
+
+	PortalIPInterval time.Duration
+	EndpointInterval time.Duration
+
+	PublicIP net.IP
+
+	ServiceIP         net.IP
+	ServicePort       int
+	PublicServicePort int
+
+	ReadOnlyServiceIP         net.IP
+	ReadOnlyServicePort       int
+	PublicReadOnlyServicePort int
+
+	runner *util.Runner
+}
+
+// Start begins the core controller loops that must exist for bootstrapping
+// a cluster.
+func (c *Controller) Start() {
+	if c.runner != nil {
+		return
+	}
+
+	repair := servicecontroller.NewRepair(c.PortalIPInterval, c.ServiceRegistry, c.PortalNet, c.ServiceIPRegistry)
+
+	// run all of the controllers once prior to returning from Start.
+	if err := repair.RunOnce(); err != nil {
+		glog.Errorf("Unable to perform initial IP allocation check: %v", err)
+	}
+	if err := c.UpdateKubernetesService(); err != nil {
+		glog.Errorf("Unable to perform initial Kubernetes service initialization: %v", err)
+	}
+	if err := c.UpdateKubernetesROService(); err != nil {
+		glog.Errorf("Unable to perform initial Kubernetes RO service initialization: %v", err)
+	}
+
+	c.runner = util.NewRunner(c.RunKubernetesService, c.RunKubernetesROService, repair.RunUntil)
+	c.runner.Start()
+}
+
+// RunKubernetesROService periodically updates the kubernetes service
+func (c *Controller) RunKubernetesService(ch chan struct{}) {
+	util.Until(func() {
+		if err := c.UpdateKubernetesService(); err != nil {
+			util.HandleError(fmt.Errorf("unable to sync kubernetes service: %v", err))
+		}
+	}, c.EndpointInterval, ch)
+}
+
+// ServiceWriterLoop is exposed for downstream consumers of master
+func (c *Controller) UpdateKubernetesService() error {
+	// Update service & endpoint records.
+	// TODO: when it becomes possible to change this stuff,
+	// stop polling and start watching.
+	// TODO: add endpoints of all replicas, not just the elected master.
+	if err := c.CreateNamespaceIfNeeded(api.NamespaceDefault); err != nil {
+		return err
+	}
+	if c.ServiceIP != nil {
+		if err := c.CreateMasterServiceIfNeeded("kubernetes", c.ServiceIP, c.ServicePort); err != nil {
+			return err
+		}
+		if err := c.SetEndpoints("kubernetes", c.PublicIP, c.PublicServicePort); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// RunKubernetesROService periodically updates the kubernetes RO service
+func (c *Controller) RunKubernetesROService(ch chan struct{}) {
+	util.Until(func() {
+		if err := c.UpdateKubernetesROService(); err != nil {
+			util.HandleError(fmt.Errorf("unable to sync kubernetes RO service: %v", err))
+		}
+	}, c.EndpointInterval, ch)
+}
+
+// ROServiceWriterLoop is exposed for downstream consumers of master
+func (c *Controller) UpdateKubernetesROService() error {
+	// Update service & endpoint records.
+	// TODO: when it becomes possible to change this stuff,
+	// stop polling and start watching.
+	if err := c.CreateNamespaceIfNeeded(api.NamespaceDefault); err != nil {
```

Comment:
Right - my point was that they don't both need to do it indiscrimately - only one of them needs to do it indiscriminately (to ensure the namespace is recreated if it deletes it) and the other one could do it only if the service actually needs to be created  (i.e. if c.ReadOnlyServiceIP != nil).

You can ignore it, though, it's really minor.